README.md
# RemoteResource

Supports PHP 5.3.2 and up.

### Philosophy

RemoteResource adopts most of the public API of ActiveResource for Rails and enforces many of the same expectations.

### Documentation

##### Available Instance Methods

```
returns     method( args )
_____________________________________________________
boolean     updateAttributes( $attributes = array() )
null        destroy()
boolean     save()
boolean     persisted()
boolean     valid()
integer     id()
array       errors()
array       attributes()
```

##### Available Class Methods

```
returns                      method( args )
____________________________________________________________
RemoteResource               create( $attributes = array() )
RemoteResource               find( $id )
RemoteResourceCollection     where( $attributes_to_match = array() )
RemoteResourceCollection     all()
```

### Usage

##### Creating a Model

Extend RemoteResource\RemoteResource, and set static $site.
$resource_name and $plural_resource_name properties are optional, and may be set where the name of the class does not mirror the name of the remote resource.
Configuration is set on a per-resource basis, designated by the $format, $auth_type, and $credentials properties.

```php
class ProductImage extends RemoteResource\RemoteResource {
  public static $site                 = "http://example.com/api/product_images";
  public static $resource_name        = "product_image";
  public static $plural_resource_name = "product_images";
  public static $format               = "json";
  public static $auth_type            = "basic";
  public static $credentials          = "user:password";
}
```

##### Simplest Model

```php
class ProductImage extends RemoteResource\RemoteResource {
  public static $site = "http://example.com/api/product_images";
}
```

##### Attribute Assignment

Dynamically assigned attributes are added to the attributes which will be sent over on save.

```php
$product_image = new ProductImage;
$product_image->file = $file;
$product_image->name = 'rainbow cube';
$product_image->save();
```

If an attribute comes back from the remote API which has already been set on the resource instance, it will be overwritten to reflect the persisted value.

```php
$product_image = ProductImage::find(15);
echo $product_image->name; // "rainbow cube"
echo $product_image->sizes_and_urls[30]; // "https://path/to/cdn/image/sjiEFzciA.png"
```

##### Create

```php
$product_image = ProductImage::create($attributes);
```

##### Where

```php
$product_images = ProductImage::where( array('product_id' => 12) );
```

##### Find

```php
$product_image = ProductImage::find(15);
```
##### All

```php
$product_images = ProductImage::all();
```

##### Update

As with ActiveResource, the update() method is reserved for internal use. use updateAttributes() instead.

```php
$product_image->updateAttributes( array('product_id' => 12) );
```

##### Destroy

```php
$product_image->destroy();
```

##### Save

```php
$product_image->save();
```

### Expectations / Assumptions

Expectations on how the remote API is configured mirrors ActiveResource expectations. I will list those expectations here, since they aren't clearly documented for ActiveResource. __It is necessary for the remote API you are accessing to fulfill the HTTP, status codes, and JSON responses from the interface below, unless you do not plan (or do not need) to use the associated methods.__

```
method    HTTP                                 status        expected JSON response
_______________________________________________________________________________________
create    POST /remote_resource_path/          201           {'remote_resource': {}}
create    POST /remote_resource_path/          422           {'errors': []}

where     GET /remote_resource_path/           200           {'remote_resources': []}

all       GET /remote_resource_path/           200           {'remote_resources': []}

find      GET /remote_resource/path/:id        200           {'remote_resource': {}}

update    PATCH /remote_resource_path/:id      204           nil
update    PATCH /remote_resource_path/:id      422           {'errors': []}

destroy   DELETE /remote_resource_path/:id     204           nil
```

### Exceptions

Familiarize yourself with RemoteResource Exceptions. __Exceptions are evaluated in the order presented here.__

```
status_code         exception
---------------------------------------------------------------
400                 RemoteResource\Exception\BadRequest
401                 RemoteResource\Exception\UnauthorizedAccess
403                 RemoteResource\Exception\ForbiddenAccess
404                 RemoteResource\Exception\ResourceNotFound
405                 RemoteResource\Exception\MethodNotAllowed
408                 RemoteResource\Exception\RequestTimeout
409                 RemoteResource\Exception\ResourceConflict
410                 RemoteResource\Exception\ResourceGone
401..499            RemoteResource\Exception\ClientError
500..599            RemoteResource\Exception\ServerError
unknown             RemoteResource\Exception\ConnectionError
```

The response that triggered the exception is stored, and can be accessed.

```php
try {
} catch ( RemoteResource\Exception\ResourceNotFound $e) {
  $response = $e->response;
}
```

### Checking Validity

In order to determine whether or not a resource was correctly created, check the validity of the resource.

```php
$product_image = ProductImage::create($attributes);
if ( $product_image->valid() ) {
  // success
}
```

Validity is determined by whether or not errors were generated for the resource. A resource is considered valid when it has no errors. This does not necessarily mean that the resource has been persisted. Check the persisted() method for this information.

### RemoteResource\Collection

A RemoteResource\Collection is a collection of RemoteResource objects. The RemoteResource\Collection object implements the Iterator interface, and can therefore be treated as a PHP collection.

```php
$remote_resource_collection = ProductImage::all();

foreach($remote_resource_collection as $remote_resource) {
  echo get_class( $remote_resource ); // "RemoteResource\RemoteResource"
}
```

##### RemoteResource\Collection methods of interest

```
returns                      method( args )
____________________________________________________________
integer                      size() || count()
RemoteResource               first()
RemoteResource               last()
```

##### Custom Methods

RemoteResource allows you to go off of the rails of RESTful convention. Know that __responses will be returned as arrays, and not as RemoteResource objects__.

```
returns     method        ( args )
_____________________________________________________________
array       static get       ( $path, $attributes=array() )
array       static post      ( $path, $attributes=array() )

array              patch     ( $path, $attributes=array() )
array              delete    ( $path )
```

```php
// requests GET "http://example.com/product_images/5/clone/"
ProductImage::get("5/clone");

// requests POST "http://example.com/product_images/5/clone/"
ProductImage::post("5/clone", array("clone_to" => 15));

// requests PATCH "http://example.com/product_images/12/increment/"
$product_image->patch("increment");

// requests DELETE "http://example.com/product_images/12/associated/"
$product_image->delete("associated");
```

### Logging
Global logging through New Relic will be activated if new relic is configured against the application consuming RemoteResource.

### Global Config
TODO: info on Global Config here

### Changelog

View the [CHANGELOG.md](CHANGELOG.md "CHANGELOG.md")

### TODO

- README section on Formatters and Auth choices and Installation
- http_build_query may have problems with more complex query string requirements
- handle timeouts gracefully
- handle data URI conversion of files


CHANGELOG.md
#Changelog

###[0.3.1] - 2015-07-01
- explicitly set permissions on log files

###[0.3.0] - 2015-06-11
- [feature] Add api_key auth type

###[0.2.10] - 2015-06-08
- [bug] GlobalConfig setLogPath takes log_path param

###[0.2.9] - 2015-06-08
- [logging] Non-error logging now logs to the server.
- [feature] Global Configuration options now settable: app_name, log_path

###[0.2.8] - 2015-06-08
- [logging] Also logging guzzle exceptions.

###[0.2.7] - 2015-06-08
- [logging] Added global logging through monolog out to newrelic. Only activated if newrelic is configured against the application consuming RemoteResource.

###[0.2.6] - 2015-05-14
- [bug] Connection headers were not being transformed into a map. This meant that the Guzzle Client wasnt able to process the headers properly.

###[0.2.4] - 2015-04-29
- [bug] removed the setSite method on RemoteResource. The method was problematic because it is setting a static property against RemoteResource itself, which will overwrite the path globally, affecting all sub classes.
  From now on, we should not manipulate the static properties meant to be set against the sub-classes anywhere in the RemoteResource class.

###[0.2.3] - 2015-04-14
- [bug] fixed a formatting issue where a response was being json encoded, then json encoded again within Guzzle

###[0.2.1] - 2015-04-02
- requests library replaced with guzzle
- concept of Formatters introduced
- Config and Connection APIs turned from class-level to instance-level
- configuration is now set at the RemoteResource subclass level, giving full control over payload data type and authentication concerns
- groundwork laid for easily adding additional supported configurations
- custom method support

###[0.2.0] - 2015-04-01
- php namespace pattern applied
- test data pulled out of library
- fully transitioned from codeception to phpunit
- bootstrapping for phpunit
- dependencies now all declared
- more obvious class names

###[0.1.0] - 2015-04-01
- RemoteResource and RemoteResourceCollection fully functional and fully tested


src/BasicRemoteResource.php
<?php
require_once 'Requests/Requests.php';
require_once 'src/RemoteResourceException.php';
require_once 'src/RemoteResourceConfig.php';
Requests::register_autoloader();

class BasicRemoteResource {
  // GET
  public static function get($path) {
    return self::handleResponse(Requests::get($path, self::headers()));
  }

  // POST
  public static function post($path, $attributes = array()) {
    return self::handleResponse(Requests::post($path, self::headers(), json_encode($attributes)));
  }

  // PATCH
  public static function patch($path, $attributes = array()) {
    return self::handleResponse(Requests::patch($path, self::headers(), json_encode($attributes)));
  }

  // DELETE
  public static function delete($path) {
    return self::handleResponse(Requests::delete($path, self::headers()));
  }

  public static function headers() {
    $credentials = RemoteResourceConfig::base64EncodedCredentials();

    return array(
      'Content-Type' => 'application/json',
      'Authorization' => 'Basic '.$credentials
    );
  }

  private static function handleResponse($response) {
    $decoded_body = json_decode( $response->body, true );

    if ($response->status_code >= 200 && $response->status_code < 400) {
      return $decoded_body;

    } elseif ($response->status_code == 400) {
      throw new RemoteResourceBadRequest($decoded_body);

    } elseif ($response->status_code == 401) {
      throw new RemoteResourceUnauthorizedAccess($decoded_body);

    } elseif ($response->status_code == 403) {
      throw new RemoteResourceForbiddenAccess($decoded_body);

    } elseif ($response->status_code == 404) {
      throw new RemoteResourceResourceNotFound($decoded_body);

    } elseif ($response->status_code == 405) {
      throw new RemoteResourceMethodNotAllowed($decoded_body);

    } elseif ($response->status_code == 409) {
      throw new RemoteResourceResourceConflict($decoded_body);

    } elseif ($response->status_code == 410) {
      throw new RemoteResourceResourceGone($decoded_body);

    } elseif ($response->status_code == 422) {
      throw new RemoteResourceResourceInvalid($decoded_body);

    } elseif ($response->status_code >= 401 && $response->status_code < 500) {
      throw new RemoteResourceClientError($decoded_body);

    } elseif ($response->status_code >= 500 && $response->status_code < 600) {
      throw new RemoteResourceServerError($decoded_body);

    } else {
      throw new RemoteResourceConnectionError($decoded_body, "Unknown response code");
    }
  }
}
